docs/use_cases/scaling_rag_for_production.md

```diff
@@ -0,0 +1,641 @@
+                    print(f"Error with future for {url}: {e}")
+```
+
+Because the documentation is very large, we will only download a subset of it. We will use the documentation of LangChains Expression Language (LCEL), which consists of 28 html pages.
+
+```python
+base_domain = "python.langchain.com"
+start_url = "https://python.langchain.com/docs/expression_language/"
+folder = working_dir
+
+download_all(start_url, folder, max_workers=10)
+```
+
+Now that we have downloaded the files, we can use them to create our Ray dataset:
+
+```python
+from pathlib import Path
+
+# Ray dataset
+document_dir = Path(folder)
+ds = ray.data.from_items([{"path": path.absolute()} for path in document_dir.rglob("*.html") if not path.is_dir()])
+print(f"{ds.count()} documents")
+```
+
+Great! But there is something left to do before we can move on to the next phase of our workflow. We still need to extract the relevant text from our html files and clean up all the html syntax. For this, we will import BeautifulSoup to parse the files and find relevant html tags.
+
+```python
+from bs4 import BeautifulSoup, NavigableString
+
+def extract_text_from_element(element):
+    texts = []
+    for elem in element.descendants:
+        if isinstance(elem, NavigableString):
+            text = elem.strip()
+            if text:
+                texts.append(text)
+    return "\n".join(texts)
+
+def extract_main_content(record):
+    with open(record["path"], "r", encoding="utf-8") as html_file:
+        soup = BeautifulSoup(html_file, "html.parser")
+
+    main_content = soup.find(['main', 'article'])  # Add any other tags or class_="some-class-name" here
+    if main_content:
+        text = extract_text_from_element(main_content)
+    else:
+        text = "No main content found."
+
+    path = record["path"]
+    return {"path": path, "text": text}
+
+```
+
+We can now use this extraction process by utilizing Ray’s map() function. This let’s us run multiple processes in parallel.
+
+```python
+# Extract content
+content_ds = ds.map(extract_main_content)
+content_ds.count()
+
+```
+
+Awesome, this will be our dataset. Ray Datasets are optimized for performance at scale - which will make productionizing our system easier for us. Having to manually adjust code while your application grows can get very costly and is prone to errors.
+
+### Processing the data
+
+The next three processing steps will consist of chunking, embedding and indexing our data source. Chunking is the process of splitting your documents into multiple smaller parts. Not only will this be necessary to make your data meet the LLM’s context length limits, it also helps to keep contexts specific enough to remain relevant. On the other hand, if your chunks are too small, the information retrieved might become too narrow. The exact chunk size will depend on your data, the models used and your use case. We will use a standard value here that has been used in a lot of applications.

```

Comment:
not sure if there is a plural form of context(s)?

Reply:
There is, but I could substitute "contexts" with "chunks" for better clarity.